# CLImAT
 accurate detection of copy number alteration and loss of heterozygosity in impure and aneuploid tumor samples using whole-genome sequencing data

------------------------------
Introduction
###########
We presents CLImAT, a bioinformatic tool for identification of genome-wide aberrations from tumor samples using whole-genome sequencing data. Without requiring a matched normal sample, CLImAT takes integrated analysis of read count and allele frequency derived from sequenced tumor samples, and provides extensive data processing procedures including GC-content and mappability correction of read count and quantile nor-malization of B allele frequency. CLImAT can accurately identify copy number alterations and loss of heterozygosity even for highly contaminated tumor samples with aneuploidy.

------------------------------
Recent Update
###########
- 2015/02/06: The latest version of CLImAT softwares (Version 1.2.2) is available from here. In this version, we fixed bugs encountered in aberration detection on chromosome X and Y. Running errors regarding DFExtract tool are also fixed. More details can be found from here.
- 2014/12/10: The latest version of CLImAT softwares (Version 1.2.1) is available from here. In this version, copy number alterations of chromosome X and Y are also annotated. More details can be found from here.
- 2014/12/07: Reported running errors have been fixed for DFExtract software.
- 2014/10/16: Version 1.2.0 of CLImAT softwares is available from here. We integrate Bamtools into DFExtract software to improve efficiency of data processing. Quantile normalization of BAF data is automatically performed based on selection of optimal threshold by grid search. 
- 2014/06/13: Updated SNP files are available from here.
- 2014/06/13: Version 1.1 of CLImAT softwares is available from here, the function of estimating copy number for regions between distant SNPs will not be provided since version 1.1. 
- 2013/10/08: Supplementary files of CLImAT article are available from here.

------------------------------
Usage
###########
After downloading and extracting DFExtract and CLImAT packages, there are several steps needed to run CLImAT. 
######1.Use of DFExtract.
We integrate Bamtools into DFExtract to improve efficiency of data processing since version 1.2.0.
To successfully run DFExtract you will need to obtain/create these items:
A tumor BAM file
A SNP file
A genome reference file(.fasta)
A mappability file(.bw)
SNP files can be downloaded from here, chose corresponding SNP file based on the reference genome you used.
Human genome reference (hg18, hg19) can be obtained from UCSC genome browser.
A Mappability file formatted as .bw is needed to calculate mappability scores, for reference hg18, download it from http://hgdownload.cse.ucsc.edu/goldenPath/hg18/encodeDCC/wgEncodeMapability/, for reference hg19, download it from http://hgdownload.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeMapability/, we recommend that you download file wgEncodeCrgMapabilityAlign36mer.bw.
You should now have all the files needed to run DFExtract.
Usage:
./DFExtract [options] <BAM file>
|Options	|Description	|Defualt |value|
|-h, --help	|give help information	|--|
|-M, --male	|to indicate sample from a male	|--|
|-w, --window	|set the size of windows	|1000|
|-Q, --baseQ	|threshold value for base quality	|10|
|-q, --mapQ	|threshold value for mapping quality	|20|
|-d, --minDepth	|minimum read depth for a position to be considered	|10|
|-s, --snp	|SNP file	|--|
|-r, --reference	|genome reference file(.fasta)	|--|
|-m, --mappability	|mappability file(.bw)	|--|
|-o, --output	|output file	|--|
Example:
./DFExtract -M -w 1000 -Q 10 -q 20 -d 5 -s hg18_snp130_1based.txt -r Homo_sapiens_assembly18.fasta -m wgEncodeCrgMapabilityAlign36mer.bw tumor.bam > tumor.out
./DFExtract -w 1000 -Q 10 -q 20 -d 5 -s hg18_snp130_1based.txt -r Homo_sapiens_assembly18.fasta -m wgEncodeCrgMapabilityAlign36mer.bw -o tumor.out tumor.bam

######2.Use of CLImAT.
If you have Matlab installed in your machine, you are golden. Run CLImAT according to the following instructions.
Fitst, use CLImAT() function to perform batch annotation for multiple samples, then use CLImAT_plot() function to plot the results.
Usage:
CLImAT(dataDir, outputDir, genotypeDir, configFile)
CLImAT_plot(dataDir, outputDir, plotDir);

Arguments	Description
dataDir	The directory containing files generated by DFExtract
outputDir	The directory to save CLImAT results
genotypeDir	The directory to save genotypes of SNPs
plotDir	The directory to save the plots of every tumor sample
configFile	The configuration file containing user-defined parameters
Example:
CLImAT('./data', './results', './genotype', 'config.txt')
CLImAT_plot('./data', './results', './plots')

An example of the config file can be found from here.
If Matlab is not installed in your machine, verify the MATLAB Compiler Runtime (MCR) is installed and ensure you have installed version 8.1 (R2013a). The MCR can be download from MathWorks Web site. We included a shell script named "run_CLImAT.sh" in "executables" directory in the CLImAT package, users should use this shell script to run CLImAT. To use the shell script, first make it executable using following command:
chmod +x run_CLImAT.sh CLImAT CLImAT_plot

######Usage:
./run_CLImAT.sh <MCRDIR> <dataDir> <outputDir> <genotypeDir> <plotDir> <configFile>

The parameter "MCRDIR" is the directory where version 8.1 of MCR is installed.

------------------------------
Installation
###########
1.Installation of DFExtract.
DFExtract works on Unix based systems, the latest version of DFExtract can be downloaded from here. A C++ compiler, e.g. GCC, and cmake are required to compile the DFExtract package. To build binary, do as follows:
tar xfzv DFExtract.tar.gz
cd DFExtract
cmake .
make

2.Running of CLImAT.
<1>Requirement
To run CLImAT, MATLAB or MATLAB Compiler Runtime (MCR) needs to be installed on userâ€™s computer. CLImAT totally takes ~80 minutes and uses 4GB RAM when running a 250GB tumor sample.
<2>Run CLImAT
The latest version of CLImAT can be downloaded from here.
The CLImAT package can be used after extracted from archive without installation. Details of usage can be found from here 

------------------------------
Input files
###########
1.depth file
The format of the depth file generated by DFExtract is as follows:
Chr	Position	Bcount	Tcount	ReadCount	GC-content	Mappability
<table>
<tr><td>1</td><td>260</td><td>0</td><td>14</td><td>158</td><td>0.629</td><td>0.184346</td></tr>
<tr><td>1</td><td>519</td><td>0</td><td>17</td><td>152</td><td>0.629</td><td>0.187481</td></tr>
<tr><td>1</td><td>2942</td><td>0</td><td>12</td><td>48</td><td>0.596</td><td>0.1627</td></tr>
<tr><td>1</td><td>4793</td><td>16</td><td>21</td><td>70</td><td>0.62</td><td>0.169317</td></tr>
</table>
2.config file
Description of all parameters:
Parameters	Description	Default value<br>
minDepth	minimum read depth for a position to be considered	10<br>
maxDepth	maximum read depth for a position to be considered	300<br>
minGC	minimum GC-content for a position to be considered	0<br>
minMapScore	minimum mappability score for a position to be considered	0<br>
maxMapScore	maximum mappability score for a position to be considered	0.98<br>

An example of the config file is as follows:
     minDepth = 10
     maxDepth = 300
     minGC = 0
     minMapScore = 0
     maxMapScore = 0.98
     
------------------------------
Reference
###########
Zhenhua Yu, Yuanning Liu, Yi shen, Minghui Wang, Ao Li. CLImAT: accurate detection of copy number alteration and loss of heterozygosity in impure and aneuploid tumor samples using whole-genome sequencing data.

------------------------------
Contact
###########
Ao Li, Associate Professor, Department of Electronic Science and Technology, USTC (aoli at ustc dot edu dot cn )
Zhenhua Yu, Graduate Student, Department of Electronic Science and Technology, USTC (yzh163 at mail dot ustc dot edu dot cn )
